packages/dds/map/src/directory.ts

```diff
@@ -1529,8 +1539,27 @@ class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirec
      * @param op - The message if from a remote delete, or null if from a local delete
      */
     private deleteSubDirectoryCore(subdirName: string, local: boolean) {
-        // This should make the subdirectory structure unreachable so it can be GC'd and won't appear in snapshots
-        // Might want to consider cleaning out the structure more exhaustively though?
-        return this._subdirectories.delete(subdirName);
+        if (this.hasSubDirectory(subdirName)) {
+            const previousValue = this.getSubDirectory(subdirName);
+
+            // This should make the subdirectory structure unreachable so it can be GC'd and won't appear in snapshots
+            // Might want to consider cleaning out the structure more exhaustively though?
+            const successfullyRemoved = this._subdirectories.delete(subdirName);
+            if (successfullyRemoved) {
+                const event: IDirectoryValueChanged = {
+                    key: subdirName,
+                    path: this.absolutePath,
+                    previousValue,
+                };
+                this.directory.emit("valueChanged", event, local, this.directory);
```

Comment:
This is continuation of the same thought:  Listener can get a value of the key (let's say string) or some complex non-jsonable object (subdirectory) as a value for previousValue.

It might be too late, but I'd personally prefer to have valueChanged / subDirectoryCreated / subDirectoryDeleted events only (no duplication in the form of "containedValueChanged") and either
- have different objects to register for non-recursive and recursive notifications
- Add boolean to specify if registration is for recursive or not changes (this has some limitation - only last registration for given listener would stay active, in case caller registered same listener for recursive & non-recursive events).

Reply:
Agreed that subDirectoryCreated/subDirectoryDeleted would be better here.  Otherwise all customers must put a conditional in their handler to figure out what they're working with, so better served by a more-specific event name.  This will also avoid the breaking change -- as-is this will break customers who don't expect subdirectories to trigger these existing events.

Reply:
I see your point that valueChanged may not be suited for create/deleteSubDirectory. Previously I had conceptualized subdirectories as a specialized form of a key, but looking at the implementation more I can tell they are handled entirely differently.

I agree with the sentiment that valueChanged and containedValueChanged feel redundant and awkward. I think containedValueChanged is unique in that you can subscribe to events of a specific directory and not receive the events from the root directory every time, but something about the duplicated event feels off. We don't expect to use containedValueChanged for our needs.